src/Compilers/Core/Portable/CodeAnalysisEventSource.cs

```diff
@@ -65,35 +62,56 @@ internal void StopSingleGeneratorRunTime(string generatorName, string assemblyPa
         internal void NodeTransform(int nodeHashCode, string name, string tableType, int previousTable, string previousTableContent, int newTable, string newTableContent, int input1, int input2)
         {
+            Span<EventData> data = stackalloc EventData[]
+            {
+                GetEventDataForInt(nodeHashCode),
+                GetEventDataForString(name),
+                GetEventDataForString(tableType),
+                GetEventDataForInt(previousTable),
+                GetEventDataForString(previousTableContent),
+                GetEventDataForInt(newTable),
+                GetEventDataForString(newTableContent),
+                GetEventDataForInt(input1),
+                GetEventDataForInt(input2),
+            };
+
             unsafe
             {
-                fixed (char* nameBytes = name)
-                fixed (char* tableTypeBytes = tableType)
-                fixed (char* previousTableContentBytes = previousTableContent)
-                fixed (char* newTableContentBytes = newTableContent)
+                fixed (EventSource.EventData* dataPtr = data)
                 {
-                    EventSource.EventData* data = stackalloc EventSource.EventData[9];
-                    data[0].DataPointer = (IntPtr)(&nodeHashCode);
-                    data[0].Size = 4;
-                    data[1].DataPointer = (IntPtr)nameBytes;
-                    data[1].Size = ((name.Length + 1) * 2);
-                    data[2].DataPointer = (IntPtr)tableTypeBytes;
-                    data[2].Size = ((tableType.Length + 1) * 2);
-                    data[3].DataPointer = (IntPtr)(&previousTable);
-                    data[3].Size = 4;
-                    data[4].DataPointer = (IntPtr)previousTableContentBytes;
-                    data[4].Size = ((previousTableContent.Length + 1) * 2);
-                    data[5].DataPointer = (IntPtr)(&newTable);
-                    data[5].Size = 4;
-                    data[6].DataPointer = (IntPtr)newTableContentBytes;
-                    data[6].Size = ((newTableContent.Length + 1) * 2);
-                    data[7].DataPointer = (IntPtr)(&input1);
-                    data[7].Size = 4;
-                    data[8].DataPointer = (IntPtr)(&input2);
-                    data[8].Size = 4;
-                    WriteEventCore(6, 9, data);
+                    WriteEventCore(eventId: 6, data.Length, dataPtr);
                 }
             }
         }
+
+        private static unsafe EventData GetEventDataForString(string value)
+        {
+            fixed (char* ptr = value)
+            {
+                return new EventData()
+                {
+                    DataPointer = (IntPtr)ptr,
+                    Size = (value.Length + 1) * sizeof(char),
+                };
+            }
+        }
+
+        private static EventData GetEventDataForInt(int value)
```

Comment:
1. Use the runtime type for the method name
2. Caller must pass in the pointer

```suggestion
        private static EventData GetEventDataForInt32(int* value)
```